### ISSUE TITLE
Make background fetch and caching more robust by separating different responses.

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#1664]
When I turn off the background fetch of cache the front page renders completely. But when it's enabled all regions but the main disappear.

I'm on a fresh 1.3.2 install, copy of Bartik with minor changes.

I'm surprised no one else has reported this so perhaps it's just me. It's odd that it's only on the front page and that background fetch seems to prevent a full render.

**Comment by herbdool** [comment]
I should add that the initial rendering is fine, but it seems that when the front page expires it only gets partially rendered again. I've only noticed when visiting the front after 5+ minutes (when cache is set to expire)

**Comment by klonos** [comment]
Can you please upgrade to the latest 1.3.3 version report back if the issue still happens with that version?

**Comment by herbdool** [comment]
Done. It's now on 1.3.3 and background fetch is enabled. I can't replicate the behaviour. I took a look to see what was different in 1.3.3 and perhaps it's this snippet in core/modules/system/system.module which is preventing full cache clears on each cron:

```
  // Delete expired cache entries.
  // Avoid invoking hook_flush_cashes() on every cron run because some modules
  // use this hook to perform expensive rebuilding operations (which are only
  // designed to happen on full cache clears), rather than just returning a
  // list of cache tables to be cleared.
  $cache_object = cache_get('system_cache_tables');
  if (empty($cache_object)) {
    $core = array('cache', 'path', 'filter', 'page', 'menu');
    $cache_bins = array_merge(module_invoke_all('flush_caches'), $core);
    cache_set('system_cache_tables', $cache_bins);
  }
  else {
    $cache_bins = $cache_object->data;
  }
```

Anyway I'm going to keep an eye on it.

**Comment by klonos** [comment]
Great! Glad that the upgrade to 1.3.3 seems to have solved this. Lets close this issue then and you can re-open if you notice it happening again.

**Comment by herbdool** [comment]
Spoke too soon. It's broken again.

**Comment by klonos** [comment]
Is this a live site or local dev?

**Comment by herbdool** [comment]
It's a live site so I won't play around with it too much. I think I'll create a test site on Pantheon and see if I can recreate the issue.

**Comment by klonos** [comment]
I just enabled caching for anonymous users on my live site and don't see anything strange happening with the block rendering. ...but then again, I have a fairly simple site with a simple layout and only a few blocks (the default ones more or less).

**Comment by herbdool** [comment]
Mine is a blog with thousands of entries but a really simple layout with a few simple blocks. I'm not sure if the number makes any difference. I had a view for the front but then switched to default /node which made no difference. Only turning off background fetch seems to prevent issue.

**Comment by herbdool** [comment]
I've created a test site on Pantheon, created a few hundred nodes and haven't been able to replicate it -- background fetch is on, front page regions are rendering. On my live site with background fetch enabled, eventually the header, sidebar and footer regions disappear, which I suspect is happening on a cache refresh.

I was incorrect earlier when I said it was a fresh install. It's an upgrade from Drupal 7 but I had used a fresh Drupal 7 site and was only using it so I could use Migrate. So I didn't configure any layouts, etc and kept it really bare except for imported nodes before an upgrade to Backdrop.

**Comment by klonos** [comment]
Perhaps take a look at the watchdog entries for any clues? Also any errors in the console?

I honestly am running out of ideas and cannot help any further. Anyone else has any thoughts about troubleshooting this further?

**Comment by serundeputy** [comment]
I think that:

 * `tail`'ing the apache logs 
 * keeping an eye on cpu/ram possibly through `htop`

put any results here.

Could provide some additional trouble shooting info here; sounds to me like your server might be exhausting all of its RAM during the request(s).

If you use a view with pagination (load say 10 nodes at a time) does that make any difference?

~Geoff

**Comment by herbdool** [comment]
Thanks @serundeputy -- I was thinking that it might have something to do with the size of the front. It's not clear to me why the problem only arises when rebuilding the cache in the background. I'm going to take a look at the cpu/ram and see.

**Comment by quicksketch** [comment]
Thanks @herbdool for reporting this issue and providing so much useful information.

> But when it's enabled all regions but the main disappear.

We've seen this intermittently on backdropcms.org itself. In total, I think we've seen it 3 or 4 times over the life of the site. A few times on the home page and once on the /modules page (which is a view with a layout configuration override at its path). The situation was exactly the same as described, it was as though the Layout were not rendered at all but the main page content was still shown.

> It's not clear to me why the problem only arises when rebuilding the cache in the background.

Yeah this is odd to me as well. In the times that this has happened on backdropcms.org however, it was *prior* to the release of 1.3.0, which included the background fetch feature. So there may be a chance that this issue is not related to background fetch at all. However, the problem has been *extremely* intermittent. Perhaps background fetch increases the probability that such a problem occurs.

This seems like a critical bug to me, a site without its blocks is a very broken site. Any more help on consistently reproducing this issue or guesses at solving it are most welcome.

**Comment by quicksketch** [comment]
> I was thinking that it might have something to do with the size of the front.

What is the path configured to be your front page? (e.g. `/node`, `node/1`, `/front`, etc.) The only two places I've seen this bug are on pages that have a layout overriding an existing path, rather than on layouts that provide their own menu item. On backdropcms.org, the homepage that exhibited this problem was `node/1` and the modules page that exhibited it was `modules`, which is a view page. In both cases, that "main page content" was the only thing shown, with all the blocks missing. So using a layout on an overridden path may be one prerequisite for reproducing this bug.

**Comment by dboulet** [comment]
Wondering if this bug could be related to https://github.com/backdrop/backdrop-issues/issues/1491.

**Comment by quicksketch** [comment]
> Wondering if this bug could be related to #1491.

Yes! I think you may have it @dboulet. I was thinking it might be something related to a particular type of request. Then once cached, the bad copy could end up getting served to all users. Let's solve that issue and see if it corrects this one by association.

**Comment by herbdool** [comment]
> What is the path configured to be your front page?

I've had it happen on a view with a custom path set to the homepage as well as /node. Currently I've got it on /node and turned background caching off. I haven't had a chance to test more combos.

**Comment by klonos** [comment]
@herbdool were you experiencing this bug while viewing on IE or another browser though?

**Comment by dboulet** [comment]
>I was thinking it might be something related to a particular type of request. Then once cached, the bad copy could end up getting served to all users.

If that’s true, then this might have revealed a flaw in the page caching logic, as well. Responses to HTML and non-HTML requests shouldn’t be sharing a cache. Does that mean that we would also see this bug if the initial cached page content was in response to an AJAX request?

**Comment by herbdool** [comment]
@klonos I tested in just about everything but IE. Was broken in ipad Safari, android Firefox, Chrome, Linux Chrome, Firefox. So doesn't seem browser specific to me.

**Comment by quicksketch** [comment]
> Does that mean that we would also see this bug if the initial cached page content was in response to an AJAX request?

Seems that it may be possible to do this if you crafted a special request that had the right headers. However in the case of Backdrop's AJAX requests to itself (through ajax.js), they are always POST requests, so they are not cacheable in the first place.

> Was broken in ipad Safari, android Firefox, Chrome, Linux Chrome, Firefox. So doesn't seem browser specific to me.

As long as the cache gets set by a misbehaving browser, it could get used by any other browser.

@dboulet is right that we may have a larger issue here in that our page cache can overlap different `Accept` headers that may have different responses. We can do a simple fix as suggested in https://github.com/backdrop/backdrop-issues/issues/1491 that should solve the problem in all "normal" situations, but a more robust caching fix should also be implemented to separate these different responses.

**Comment by klonos** [comment]
> ...but a more robust caching fix should also be implemented to separate these different responses.

Better title then.

**Comment by herbdool** [comment]
Further testing: I did a simple test. I set front page view to be only three nodes (previously was ten) and turned background caching back on. Now I don't see the issue reappear.

This suggests that it could be timing out before it's able to properly cache the front?

By the way, for some reason it would not allow me to set anything other than ten nodes per page on the site information admin page. Bug?

**Comment by gifad** [comment]
> set anything other than ten nodes per page

missing
```
  config('system.core')
…
    ->set('default_nodes_main', $form_state['values']['default_nodes_main'])
…
```
in core/modules/system/system.admin.inc line ~1439

**Comment by herbdool** [comment]
An update: it took a day but I finally saw the error again. With just 3 nodes on front page and background caching on.

**Comment by quicksketch** [comment]
I'm fairly certain that https://github.com/backdrop/backdrop-issues/issues/1491 will solve the immediate issue. I don't think background cache is directly related to this issue. To easily reproduce:

- Install Backdrop. Turn on the page cache with a 5 minute lifetime.
- In a new anonymous session, visit the site in IE8 to set the page cache.
- Now visit the site as any other browser for the next 5 minutes as anonymous.

This will happen regardless of whether the Background Fetch is turned on or off. That problem overall is solved by the above issue. It will fix requests coming in from IE8 and identify them as HTML page requests.

While the problem will be solved for IE8, we still have deeper issues that the caches can get mixed up based on the type of content being requested. If a JSON version of a page is requested, we don't want that getting mixed up with the HTML version of the page. Right now that's unlikely to happen because all our JSON requests are done by POST, which prevents cache corruption, but isn't very efficient from a caching perspective.

We currently determine the page type based on the "Accept" HTTP header. As pointed out in [this article](http://www.newmediacampaigns.com/blog/browser-rest-http-accept-headers) the Vary headers are very unreliable across different browsers. Even if we normalized the Accept header to just a set of expected values (e.g. `text/html`, `text/javascript`, `application/vnd.backdrop-dialog`) we could only store that normalization in Backdrop's own page cache. External page caches like Varnish or nginx wouldn't have any idea about this normalization and the difference between a page requested as HTML versus one that was intended to be a dialog, so we'd just end up with cache mix-ups at a different layer.

Overall, the Drupal issue at https://www.drupal.org/node/2364011 sums it up much better. Ultimately, D8 went with query strings to identify the type of page request. If we adjusted `ajax.js` in Backdrop to use query strings instead of the `Accept` header, we wouldn't have to adjust the caching layer at all (and it will work with Varnish, nginx, CloudFlare, Akamai, etc.) We just need to decide if query strings seem okay and what to name them. e.g. `_format=backdrop_dialog`, `_type=backdrop_ajax` or whatever we want to come up with. D8 used `_format=json` and a second parameter for `_drupal_wrapper=dialog`, which I'm not sure about separating those two.

**Comment by herbdool** [comment]
I upgraded to 1.3.4 and went through the steps (tried to set cache with IE8), albeit on an existing site. I haven't been able to recreate the issue I had before. So perhaps the PR fixed this. I'll see if any problems reappear.

**Comment by herbdool** [comment]
Still zero issues since the upgrade. All working as expected. Can we close this issue?

**Comment by BWPanda** [comment]
> Can we close this issue?

Looks like it.
